File: day24/electromagnet_moat.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);
use List::Util qw(sum);

my %has;
my %weight;
while (<>) {
    chomp;
    my ($p0, $p1) = split '/';
    push @{$has{$p0}}, $_;
    push @{$has{$p1}}, $_ unless $p0 == $p1;
    $weight{$_} = $p0 + $p1;
}

my $maxw = 0;
my $maxb;

make_bridge(0, []);

sub make_bridge($target, $bridge) {
    my %used = map { $_ => 1 } @$bridge;
    my @next = grep { !$used{$_} } @{$has{$target}};
    if (@next) {
        for my $port (@next) {
            my @b2 = (@$bridge, $port);
            my $t2 = other_side($port, $target);
            make_bridge($t2, \@b2);
        }
    } else {
        my $w = bridge_weight($bridge);
        if ($w > $maxw) {
            $maxw = $w;
            $maxb = join '--', @$bridge;
            say "$maxw\t$maxb";
        }
    }
}

sub other_side($port, $side1) {
    my ($p0, $p1) = split '/', $port;
    return ($p0 == $side1) ? $p1 : $p0;
}

sub bridge_weight($bridge) {
    return sum(map { $weight{$_} } @$bridge);
}

File: day24/electromagnet_moat.py
#!/usr/bin/env python3
from sys import argv

has = {}
weight = {}
maxw = 0
maxb = ''

def make_bridge(target, bridge):
    global maxw
    global maxb
    
    used = { port for port in bridge }
    ports = [ port for port in has[target] if port not in used ]
    if (len(ports) > 0):
        for port in ports:
            b2 = []
            b2 += bridge
            b2.append(port)
            t2 = other_side(port, target)
            make_bridge(t2, b2)
    else:
        w = bridge_weight(bridge)
        if w > maxw:
            maxw = w
            maxb = '--'.join(bridge)
            print("{}\t{}".format(maxw, maxb))

def other_side(port, side1):
    p0,p1 = [ int(x) for x in port.split('/') ]
    return p1 if p0 == side1 else p0

def bridge_weight(bridge):
    return sum([ weight[port] for port in bridge ])        

# parse the input
filename = argv[1]
with open(filename) as f:
    for line in f:
        line = line.rstrip()

        p0,p1 = [ int(x) for x in line.split('/') ]
        if p0 not in has:
            has[p0] = [line]
        else:
            has[p0].append(line)
        if p0 != p1:
            if p1 not in has:
                has[p1] = [line]
            else:
                has[p1].append(line)
        weight[line] = p0 + p1

make_bridge(0, [])
